tools/flowc/manipulation/flatten_unions.flow

```diff
@@ -303,3 +307,86 @@ fiResolveNatives(program : FiProgram, native_is_implemented: (string) -> bool) -
 		)
 	);
 }
+
+fiAddSuperUnions(program : FiProgram) -> FiProgram {
+	names = program.names;
+	all_unions = getTreeValues(names.unions);
+
+	gather_super_unions = \acc0, supers -> foldTree(supers, acc0, \str, unions, acc ->
+		if (length(unions) <= 1) {
+			// A struct is cointained in a single / none union - it is its super union.
+			acc;
+		} else {
+			// A set of all types in a super union for a struct 'str'
+			super_names = set2array(fold(unions, makeSet(), \ac, un ->
+				fold(un.typenames, ac, \a, tn -> insertSet(a, tn))
+			));
+			// try to find an existing superunion among a program unions
+			switch (find(all_unions, \un -> forall(super_names, \tn -> contains(un.typenames, tn)))) {
+				Some(super_union): {
+					// no need to create a new super union
+					acc;
+				}
+				None(): {
+					// try to find an existing superunion among a constructed new super unions
+					switch (findList(acc, \p -> forall(super_names, \tn -> contains(p.first.typenames, tn)))) {
+						Some(p): {
+							// no need to create a new super union
+							acc;
+						}
+						None(): {
+							// A super union for a current struct doesn't exist - let's create it and add it
+							super_union = FiTypeUnion(
+								"super_" + superglue(super_names, \tn -> tn.name, "_"),
+								set2array(fold(super_names, makeSet(), \ac, tn ->
+									fold(tn.typeparameters, ac, \a, tp -> insertSet(ac, tp))
+								)),
+								super_names,
+								-1
+							);
+							Cons(Pair(super_union, str), acc);
+						}
+					}
+				}
+			}
+		}
+	);
+
+	// Gather all unions, where a struct belongs to
+	struct2unions = fold(all_unions, makeTree(), \acc, union ->
+		fold(union.typenames, acc, \ac, tn ->
+			setTree(ac, tn.name, insertSet(lookupTreeDef(ac, tn.name, makeSet()), union))
+		)
+	);
+	// List of pairs: a super union and a corresponding struct name
+	struct_super_unions = gather_super_unions(makeList(), mapTree(struct2unions, set2array));
```

Comment:
we don't use this naming convention

Reply:
What naming convention do you mean here? I have a clear meaning of what is 'struct super union' here - it's any union, which contains a struct, so I use it as a name of a variable.

Reply:
we use camelCase , and not snake_case

Reply:
@dmitry-vlasov  FYI https://github.com/area9innovation/flow9/blob/master/doc/flow.markdown#coding-conventions